src/test/java/com/aws/greengrass/clientdevices/auth/connectivity/ConnectivityInformationTest.java

```diff
@@ -0,0 +1,200 @@
+        GetConnectivityInfoResponse getConnectivityInfoResponse = GetConnectivityInfoResponse.builder().build();
+        doReturn(getConnectivityInfoResponse).when(greengrassV2DataClient)
+                .getConnectivityInfo(any(GetConnectivityInfoRequest.class));
+
+        List<ConnectivityInfo> connectivityInfos = connectivityInformation.getConnectivityInfo();
+        verify(greengrassV2DataClient, times(1))
+                .getConnectivityInfo(any(GetConnectivityInfoRequest.class));
+        assertThat(connectivityInfos, is(empty()));
+    }
+
+    @Test
+    void GIVEN_cloudThrowValidationException_WHEN_get_connectivity_info_THEN_no_connectivity_info_returned(
+            ExtensionContext context) {
+        ignoreExceptionOfType(context, ValidationException.class);
+        when(greengrassV2DataClient.getConnectivityInfo(any(GetConnectivityInfoRequest.class)))
+                .thenThrow(ValidationException.class);
+
+        assertThat(connectivityInformation.getConnectivityInfo(), is(empty()));
+    }
+
+    @Test
+    void GIVEN_cached_connectivity_info_WHEN_get_cached_connectivity_info_THEN_connectivity_info_returned() {
+        ConnectivityInfo connectivityInfo = ConnectivityInfo.builder().hostAddress("172.8.8.10")
+                .metadata("").id("172.8.8.10").portNumber(8883).build();
+        ConnectivityInfo connectivityInfo1 = ConnectivityInfo.builder().hostAddress("localhost")
+                .metadata("").id("localhost").portNumber(8883).build();
+        GetConnectivityInfoResponse getConnectivityInfoResponse = GetConnectivityInfoResponse.builder()
+                .connectivityInfo(Arrays.asList(connectivityInfo, connectivityInfo1)).build();
+        doReturn(getConnectivityInfoResponse).when(greengrassV2DataClient)
+                .getConnectivityInfo(any(GetConnectivityInfoRequest.class));
+
+        connectivityInformation.getConnectivityInfo();
+        List<String> connectivityInfos = connectivityInformation.getCachedHostAddresses();
+        assertThat(connectivityInfos, containsInAnyOrder("172.8.8.10", "localhost"));
+    }
+
+    @Test
+    void GIVEN_missingConnectivityInfo_WHEN_getConnectivityInformation_THEN_returnEmptySet() {
+        Set<HostAddress> connectivityInfo = connectivityInformation.getAggregatedConnectivityInformation();
+        assertThat(connectivityInfo, is(empty()));
+    }
+
+    @Test
+    void GIVEN_connectivityInfoFromSingleSource_WHEN_hostAddressesAddedAndRemoved_THEN_correctConnectivityInfoReturned() {
+        Set<HostAddress> sourceConnectivityInfo = Stream.of("localhost", "127.0.0.1")
+                .map(HostAddress::of)
+                .collect(Collectors.toSet());
+        Set<HostAddress> connectivityInfoSuperset = Stream.of("localhost", "127.0.0.1", "127.0.0.2")
+                .map(HostAddress::of)
+                .collect(Collectors.toSet());
+
+        connectivityInformation.updateConnectivityInformationForSource("source", sourceConnectivityInfo);
```

Comment:
Same comment as the one above. Here is one article that I love that talks about the [testing trophy](https://kentcdodds.com/blog/the-testing-trophy-and-testing-classifications) instead of the testing pyramid